### ISSUE TITLE
Eval bug: Gemma 4 & Qwen 3.5 full prompt reprocessing from system prompt in OpenCode and Pi Coding Agent but only with llama.ccp server as backend.

### ISSUE THREAD

**Issue by vevi33** [ggml-org/llama.cpp#21912]
### Name and Version

[b8789](https://github.com/ggml-org/llama.cpp/releases/tag/b8789)

### Operating systems

Windows

### GGML backends

HIP

### Hardware

GPU: Rx 7800 XT
CPU: R7 7700X
Happens with Vulkan as well.

### Models

All Gemma 4 Unsloth models
All Qwen 3.5 Unsloth models

### Problem description & steps to reproduce

I have an issue where gemma-4 is sometimes reprocessing the whole prompt after the long assistant message and many tool calls. It happens with OpenCode and Pi Coding Agent as well.
LLama.CCP HIP and VULKAN backend.
Oddly enough it is not happening with LM studio as backend.

So I ask a long task --- Model starts to think, tool call (many SWA checkpoints saved) ---- Model writes final answer (no full reprocessing yet) ---- BOOM, when I write a new message everything is reprocessed after system prompt.

I attached a log about it.
I run the server with:

@echo off
title Gemma 4 26B - Llama Server

D:\Apps\llamahip\llama-server.exe ^
  -m "G:\Ai\AiModels\lmstudio-community\unsloth\Gemma\gemma-4-26B-A4B-it-UD-Q6_K_XL.gguf" ^
  --flash-attn on ^
  --ctx-size 100000 ^
  --n-gpu-layers 31 ^
  --threads 8 ^
  --parallel 1 ^
  --no-mmap ^
  --mlock ^
  --cache-ram 8196 ^
  --ctx-checkpoints 8 ^
  --temp 0.55 ^
  --min-p 0.05 ^
  --n-cpu-moe 8 ^
  --top-p 0.95 ^
  --top-k 30 ^
  --alias "gemma@q6_k_xl" ^
  --jinja ^
  --reasoning on

### Relevant log output

slot      release: id  0 | task 4647 | stop processing: n_tokens = 13387, truncated = 0
srv  update_slots: all slots are idle
srv  params_from_: Chat format: peg-gemma4
slot get_availabl: id  0 | task -1 | selected slot by LCP similarity, sim_best = 0.122 (> 0.100 thold), f_keep = 0.113
srv  get_availabl: updating prompt cache
srv   prompt_save:  - saving prompt with length 13387, total state size = 481.749 MiB
srv          load:  - looking for better prompt, base f_keep = 0.113, sim = 0.122
srv        update:  - cache state: 2 prompts, 4020.147 MiB (limits: 8196.000 MiB, 100096 tokens, 100096 est)
srv        update:    - prompt 0000022E4256B2A0:    5662 tokens, checkpoints:  3,  1310.724 MiB
srv        update:    - prompt 0000022E6CE4E500:   13387 tokens, checkpoints:  8,  2709.423 MiB
srv  get_availabl: prompt cache update took 560.23 ms
slot launch_slot_: id  0 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> min-p -> ?xtc -> temp-ext -> dist
slot launch_slot_: id  0 | task 4896 | processing task, is_child = 0
slot update_slots: id  0 | task 4896 | new prompt, n_ctx_slot = 100096, n_keep = 0, task.n_tokens = 12378
slot update_slots: id  0 | task 4896 | n_past = 1509, slot.prompt.tokens.size() = 13387, seq_id = 0, pos_min = 12260, n_swa = 1024
slot update_slots: id  0 | task 4896 | Checking checkpoint with [11604, 13138] against 485...
slot update_slots: id  0 | task 4896 | Checking checkpoint with [11091, 12626] against 485...
slot update_slots: id  0 | task 4896 | Checking checkpoint with [7530, 9065] against 485...
slot update_slots: id  0 | task 4896 | Checking checkpoint with [7018, 8553] against 485...
slot update_slots: id  0 | task 4896 | Checking checkpoint with [812, 2160] against 485...
slot update_slots: id  0 | task 4896 | Checking checkpoint with [504, 1837] against 485...
slot update_slots: id  0 | task 4896 | Checking checkpoint with [156, 1504] against 485...
slot update_slots: id  0 | task 4896 | restored context checkpoint (pos_min = 156, pos_max = 1504, n_tokens = 1505, n_past = 1504, size = 263.493 MiB)
slot update_slots: id  0 | task 4896 | erased invalidated context checkpoint (pos_min = 504, pos_max = 1837, n_tokens = 1838, n_swa = 1024, pos_next = 1504, size = 260.563 MiB)
slot update_slots: id  0 | task 4896 | erased invalidated context checkpoint (pos_min = 812, pos_max = 2160, n_tokens = 2161, n_swa = 1024, pos_next = 1504, size = 263.493 MiB)
slot update_slots: id  0 | task 4896 | erased invalidated context checkpoint (pos_min = 7018, pos_max = 8553, n_tokens = 8554, n_swa = 1024, pos_next = 1504, size = 300.018 MiB)
slot update_slots: id  0 | task 4896 | erased invalidated context checkpoint (pos_min = 7530, pos_max = 9065, n_tokens = 9066, n_swa = 1024, pos_next = 1504, size = 300.018 MiB)
slot update_slots: id  0 | task 4896 | erased invalidated context checkpoint (pos_min = 11091, pos_max = 12626, n_tokens = 12627, n_swa = 1024, pos_next = 1504, size = 300.018 MiB)
slot update_slots: id  0 | task 4896 | erased invalidated context checkpoint (pos_min = 11604, pos_max = 13138, n_tokens = 13139, n_swa = 1024, pos_next = 1504, size = 299.823 MiB)
slot update_slots: id  0 | task 4896 | n_tokens = 1504, memory_seq_rm [1504, end)
slot update_slots: id  0 | task 4896 | prompt processing progress, n_tokens = 3552, batch.n_tokens = 2048, progress = 0.286961
slot update_slots: id  0 | task 4896 | n_tokens = 3552, memory_seq_rm [3552, end)
slot update_slots: id  0 | task 4896 | prompt processing progress, n_tokens = 5600, batch.n_tokens = 2048, progress = 0.452416
slot update_slots: id  0 | task 4896 | n_tokens = 5600, memory_seq_rm [5600, end)

I start the server with:

@echo off
title Gemma 4 26B - Llama Server

D:\Apps\llamaccprcom\llama-server.exe ^
  -m "G:\Ai\AiModels\lmstudio-community\unsloth\Gemma\gemma-4-26B-A4B-it-UD-Q6_K_XL.gguf" ^
  --flash-attn on ^
  --ctx-size 100000 ^
  --n-gpu-layers 31 ^
  --threads 8 ^
  --parallel 1 ^
  --no-mmap ^
  --mlock ^
  --cache-ram 8196 ^
  --ctx-checkpoints 8 ^
  --temp 0.55 ^
  --min-p 0.05 ^
  --n-cpu-moe 8 ^
  --top-p 0.95 ^
  --top-k 30 ^
  --alias "gemma@q6_k_xl" ^
  --jinja ^
  --reasoning on

**Comment by aldehir** [comment]
Yes, this is by design. It retains reasoning between tool calls. Once it's done and you send a message, those traces are removed from the prompt triggering prompt processing.

You can customize the template to best fit your needs. Either don't retain reasoning, sacrificing tool calling performance. Or, don't remove them and potentially degrade model performance later (or maybe not, I haven't tried).

**Comment by vevi33** [comment]
Thank you!
You were right... I fixed it with Claude. If anyone is interested, here is the fixed template. It always includes all thinking in the prompt. It doesn't trim anything. Working perfectly fine without reprocessing.

Just rename it to .jinja since I can't upload that here. :)

[GemmaJinjaAlwaysRetain.txt](https://github.com/user-attachments/files/26728067/GemmaJinjaAlwaysRetain.txt)
